pkg/operator/events/recorder_upstream.go

```diff
@@ -59,12 +126,35 @@ func (r *upstreamRecorder) Warningf(reason, messageFmt string, args ...interface
 	r.Warning(reason, fmt.Sprintf(messageFmt, args...))
 }
 
+func (r *upstreamRecorder) incrementEventsCounter(severity string) {
+	if r.involvedObjectRef == nil {
+		return
+	}
+	eventsCounterMetric.WithLabelValues(severity).Inc()
+}
+
+func (r *upstreamRecorder) isShuttingDown() bool {
+	r.shutdownMutex.RLock()
+	defer r.shutdownMutex.RUnlock()
+	return r.shuttingDown
+}
+
 // Event emits the normal type event.
 func (r *upstreamRecorder) Event(reason, message string) {
+	defer r.incrementEventsCounter(corev1.EventTypeNormal)
+	if r.isShuttingDown() {
+		r.fallbackRecorder.Event(reason, message)
+		return
+	}
 	r.eventRecorder.Event(r.involvedObjectRef, corev1.EventTypeNormal, reason, message)
 }
 
 // Warning emits the warning type event.
 func (r *upstreamRecorder) Warning(reason, message string) {
+	defer r.incrementEventsCounter(corev1.EventTypeWarning)
+	if r.isShuttingDown() {
```

Comment:
this races

Reply:
shutdown can be called after isShuttingDown left the lock. We would lose events.